@W-20342599 Password Reset uses `email` mode

hajinsuha1 opened the pull request:
<!--- Provide a short summary of your changes in the Title field above -->

# Description

<!--- A longer summary of your changes, including: a description of the issue that you’re addressing, a list of required dependencies (if applicable), and any other relevant context. -->

Added a new `mode` to the password reset configuration that allows merchants to configure the mode they want to use for all `oauth/password/reset` calls:

Now it will set the password reset mode to the following
```
# Email mode
resetPassword: {
   mode: 'email',
   landingPath: '/reset-password-landing'
},

# Callback mode
resetPassword: {
   mode: 'callback',
   callbackURI: 'https://webhook.site/e6f88064-36bc-4605-8128-a94810bd4318',
   landingPath: '/reset-password-landing'
},
```

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [x] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

**commerce-sdk-react**
- Update `getPasswordResetToken` to default locale to the one in CommerceApiProvider and pass callback_uri and idp_name only when they are defined
- Update `resetPassword` to default hint to `cross_device` and pass code_verifier only when it is defined
**pwa-kit-create-app**
- Update `default.js` and `/_app-config/index.jsx` template to use email mode by default for passwordless login and password reset.
**template-retail-react-app**
- Update password reset to use email mode by default. The mode can now be configured via default.js

# How to Test-Drive This PR

## Email mode
This tests when default.js is configured to the following:
```
resetPassword: {
   mode: 'email',
   landingPath: '/reset-password-landing'
}
```
1. Configure and start the `template-retail-react-app`
    ```
    # Checkout this branch
    git checkout W-20342599-password-reset-uses-slas-email-mode

    # Revert this commit to setup `template-retail-react-app` with private client and updates the default.js commerceApi configuraton
    git revert 1bf6452787cbc591b1494398fdcf993f0a6eca1f

    # Build commerce-sdk-react locally by building the entire app
    npm ci
    
    # Set the private client secret via env vars and start the `template-retail-react-app`
    export PWA_KIT_SLAS_CLIENT_SECRET=SECRET
    cd packages/template-retail-react-app
    npm start
    ```
1. Once the app has started, open the developer console, start the reset password flow by opening the /login page, entering your email, clicking the `Password` button, clicking `Forgot Password`, then clicking the`Reset Password` button
1. Verify an email was sent to you with the magic link (the magic link will not work as there is a bug that will be addressed in [W-20662318](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00002RkPbtYAF/view))
1. Repeat steps 2-3 for the http://localhost:3000/us/login page (verify the locale that is sent is `en-US` instead of `en-GB` in the `/password/reset` request)
1. Repeat steps 2-3 for the login from the checkout page

## Callback mode
1. Update `default.js` to
    ```
                resetPassword: {
                    mode: 'callback',
                    callbackURI: 'https://webhook.site/893aef41-13a0-41e9-8c86-95de1594ad0b',
                    landingPath: '/reset-password-landing'
                }
    ```
1. Restart the server
    ```
    npm start
    ```
1. Once the app has started, open the developer console, start the reset password flow by opening the /login page, entering your email, clicking the `Password` button, clicking `Forgot Password`, then clicking the`Reset Password` button
1. Verify SLAS sent a POST request to the webhook callback uri: https://webhook.site/893aef41-13a0-41e9-8c86-95de1594ad0b
1. Repeat steps 3-4 for the http://localhost:3000/us/login page
1. Repeat steps 3-4 for the login from the checkout page

## Backwards Compatibilty with `template-retail-react-app`
The current `template-retail-react-app` does not specify a `mode` and only specifies a `callbackURI`. Let's make sure our changes in `commerce-sdk-react` are backwards compatible!
This ensures when mode is missing but callbackURI is set in default.js, callback mode is used

1. Update `default.js` to
    ```
            resetPassword: {
                callbackURI: 'https://webhook.site/893aef41-13a0-41e9-8c86-95de1594ad0b',
                landingPath: '/reset-password-landing'
            }
    ```
1. Restart the server
    ```
    npm start
    ```
1. Once the app has started, open the developer console, start the reset password flow by opening the /login page, entering your email, clicking the `Password` button, clicking `Forgot Password`, then clicking the`Reset Password` button
1. Verify SLAS sent a POST request to the webhook callback uri: https://webhook.site/893aef41-13a0-41e9-8c86-95de1594ad0b

## pwa-kit-create-app

1. Verify projects generated using `pwa-kit-create-app` have default mode set to `email` for `resetPassword`
  ```
  # Generate the project and verify config/default.js has resetPassword.mode set to email
  node packages/pwa-kit-create-app/scripts/create-mobify-app.js --outputDir ./retail-react-app-demo-4 --preset retail-react-app-demo
  ```

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [x] Changes are covered by test cases
- [x] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [x] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

hajinsuha1 committed:
Update passwordless login configuration to default to email mode and improve documentation

hajinsuha1 committed:
Fix formatting of commented callbackURI in passwordless login configuration files

hajinsuha1 committed:
Reapply "Enable SLAS private client and passwordless login email mode, update Commerce API parameters to test with staging-001"

This reverts commit 66bcafa4d5677e691c1f55fa37987e17b71e66d3.

hajinsuha1 committed:
Refactor Auth class to conditionally include parameters in requests

- Updated callback_uri and idp_name to be included only if provided.
- Set default value for locale to use the one in the CommerceAPIProvider similar to passwordless login
- Set default value for hint to `cross_device` to improve request handling.

hajinsuha1 committed:
Update password reset hook to use new URL building utility and include mode parameter

- Refactored usePasswordReset hook to utilize buildAbsoluteUrl for constructing callbackURI.
- Updated configuration to set default mode for password reset to 'email'.
- Enhanced getPasswordResetToken function to include mode in the mutation request.

hajinsuha1 committed:
Merge branch 'W-20443849-set-default-passwordless-mode-to-email-in-hbs' into W-20342599-password-reset-uses-slas-email-mode

hajinsuha1 committed:
Update hbs to default reset password to email mode

cc-prodsec commented:
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-critical.svg) Critical | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-high.svg) High | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-medium.svg) Medium | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-low.svg) Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/c41c26a7-6c05-4a66-a632-ae4de4580dbf?refs=67cb9d81260d38196a229628bbdca4ce602ab551&source=prChecksComment) |
| :white_check_mark: | Licenses| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/c41c26a7-6c05-4a66-a632-ae4de4580dbf/license?refs=67cb9d81260d38196a229628bbdca4ce602ab551&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

hajinsuha1 committed:
Merge branch 'feature/email-otp' into W-20342599-password-reset-uses-slas-email-mode

Signed-off-by: Jinsu Ha <91205717+hajinsuha1@users.noreply.github.com>

hajinsuha1 committed:
update changelogs

hajinsuha1 committed:
Merge branch 'W-20342599-password-reset-uses-slas-email-mode' of https://github.com/SalesforceCommerceCloud/pwa-kit into W-20342599-password-reset-uses-slas-email-mode

hajinsuha1 committed:
Revert "Reapply "Enable SLAS private client and passwordless login email mode, update Commerce API parameters to test with staging-001""

This reverts commit 6d6b90f626a37b63e9763e1bfff97f7c3ce51766.

hajinsuha1 committed:
add unit tests

hajinsuha1 committed:
Update usePasswordReset hook to handle optional callbackURI configuration

- Modified the usePasswordReset hook to safely access the callbackURI property using optional chaining.
- This change ensures that the application does not break if the callbackURI is not defined in the configuration.

jeremy-jung1 reviewed (approved):
LGTM

hajinsuha1 committed:
Merge branch 'feature/email-otp' into W-20342599-password-reset-uses-slas-email-mode

Signed-off-by: Jinsu Ha <91205717+hajinsuha1@users.noreply.github.com>
